File: Protocol/Responses/AddContactResponse.cpp
#include "AddContactResponse.h"

AddContactResponse::AddContactResponse(AddContactResult result)
{
    this->result = result;
}

AddContactResult AddContactResponse::GetResult() const
{
    return this->result;
}

ResponseType AddContactResponse::GetResponseType() const
{
    return ResponseType::AddContact;
}

string AddContactResponse::ToString() const
{
    ostringstream stream;
    stream << (int)GetResponseType() << ' ' << (int)GetResult();
    return stream.str();
}

AddContactResponse *AddContactResponse::Parse(string str)
{
    istringstream stream(str);
    int responseType, result;
    stream >> responseType >> result;

    if ((ResponseType)responseType != ResponseType::AddContact)
        throw "Invalid type of request";

    return new AddContactResponse((AddContactResult)result);
}

File: Client/Protocol/Responses/AddContactResponse.cs
using System;
using System.IO;

namespace NetTalk.Protocol.Responses
{
    public enum AddContactResult
    {
        Ok,
        UserNotFound,
    };

    public class AddContactResponse : ResponseBase
    {
        public AddContactResponse(AddContactResult result)
        {
            Result = result;
        }

        public override ResponseType ResponseType { get; } = ResponseType.AddContact;
        public AddContactResult Result { get; private set; }

        public static AddContactResponse Parse(byte[] bytes)
        {
            using (MemoryStream stream = new MemoryStream(bytes))
            using (BinaryReader reader = new BinaryReader(stream))
            {
                ResponseType responseType = (ResponseType)reader.ReadInt32();
                if (responseType != ResponseType.AddContact)
                    throw new InvalidOperationException("Invalid type of response");

                AddContactResult result = (AddContactResult)reader.ReadInt32();

                return new AddContactResponse(result);
            }
        }
    }
}